core/activitystreak/activitiystreak.go

```diff
@@ -0,0 +1,249 @@
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package activitystreak
+
+import (
+	"context"
+	"sort"
+
+	"code.vegaprotocol.io/vega/core/events"
+	"code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/libs/num"
+	"code.vegaprotocol.io/vega/logging"
+	"golang.org/x/exp/maps"
+
+	vegapb "code.vegaprotocol.io/vega/protos/vega"
+	eventspb "code.vegaprotocol.io/vega/protos/vega/events/v1"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/mocks.go -package mocks code.vegaprotocol.io/vega/core/activitystreak Broker,MarketsStatsAggregator
+
+type PartyActivity struct {
+	Active                               uint64
+	Inactive                             uint64
+	RewardDistributionActivityMultiplier num.Decimal
+	RewardVestingActivityMultiplier      num.Decimal
+}
+
+func (p PartyActivity) IsActive() bool {
+	return p.Active != 0 && p.Inactive == 0
+}
+
+func (p *PartyActivity) ResetMultipliers() {
+	p.RewardDistributionActivityMultiplier = num.DecimalOne()
+	p.RewardVestingActivityMultiplier = num.DecimalOne()
+}
+
+func (p *PartyActivity) UpdateMultipliers(benefitTiers []*types.ActivityStreakBenefitTier) {
+	if !p.IsActive() {
+		// we are not active, nothing changes
+		return
+	}
+
+	// these are sorted properly already
+	for _, b := range benefitTiers {
+		if p.Active < b.MinimumActivityStreak {
+			break
+		}
+
+		p.RewardDistributionActivityMultiplier = b.RewardMultiplier
+		p.RewardVestingActivityMultiplier = b.VestingMultiplier
+	}
+}
+
+type MarketsStatsAggregator interface {
+	GetMarketStats() map[string]*types.MarketStats
+}
+
+type Broker interface {
+	SendBatch(evt []events.Event)
+}
+
+type Engine struct {
+	log *logging.Logger
+
+	marketStats     MarketsStatsAggregator
+	partiesActivity map[string]*PartyActivity
+	broker          Broker
+
+	benefitTiers                 []*types.ActivityStreakBenefitTier
+	minQuantumOpenNotionalVolume *num.Uint
+	minQuantumTradeVolume        *num.Uint
+}
+
+func New(
+	log *logging.Logger,
+	marketStats MarketsStatsAggregator,
+	broker Broker,
+) *Engine {
+	return &Engine{
+		log:             log,
+		partiesActivity: map[string]*PartyActivity{},
+		marketStats:     marketStats,
+		broker:          broker,
+	}
+}
+
+func (e *Engine) OnMinQuantumOpenNationalVolumeUpdate(
+	_ context.Context, v *num.Uint,
+) error {
+	e.minQuantumOpenNotionalVolume = v.Clone()
+	return nil
+}
+
+func (e *Engine) OnMinQuantumTradeVolumeUpdate(
+	_ context.Context, v *num.Uint,
+) error {
+	e.minQuantumTradeVolume = v.Clone()
+	return nil
+}
+
+func (e *Engine) OnBenefitTiersUpdate(
+	_ context.Context, v interface{},
+) error {
+	tiers, err := types.ActivityStreakBenefitTiersFromUntypedProto(v)
+	if err != nil {
+		return err
+	}
+
+	e.benefitTiers = tiers.Clone().Tiers
+	sort.Slice(e.benefitTiers, func(i, j int) bool {
+		return e.benefitTiers[i].MinimumActivityStreak < e.benefitTiers[j].MinimumActivityStreak
+	})
+	return nil
```

Comment:
that sounds quite weird to me to be honest, I can't think of this as being flexible, just it doesn't make sense? 
I mean how do you make it work if not sorted in here?

Reply:
@ValentinTrinque For referral programs I thought we confirmed we agreed to sort by `minimum_running_volume` - https://github.com/vegaprotocol/specs/pull/1829#discussion_r1274998494?

So in this context we would be sorting by `minimum_activity_streak`.

The maximum flexibility was referring to not requiring the `reward_factor` and `discount_factors` to be ascending when validating proposals in case we needed additional features to dissuade parties clumping together into a few large teams.

Reply:
right which I've done. so we are good.